### PR TITLE
tweak error reporting and add test of all param

### PR DESCRIPTION
**Pull request by jeremiahpslewis**

### CODE REVIEW

**Comment by jeremiahpslewis** [comment]
Further improvements:
- allows multiple user_types to be specified
- further simplifies package through refactoring

**Comment by Ironholds** [comment]
Same concerns about the granularity argument; I appreciate the user-type adding though! Further comments inline.

File: R/query.R [hunk 1]
```diff
@@ -34,33 +34,18 @@
...
-  }
-
-  # Construct parameters
-  parameters <- paste("per-article", project, ifelse(platform == "all", "all-access", platform),
-                      ifelse(user_type == "all", "all-agents", user_type), curl::curl_escape(article),
-                      "daily", start, end, sep = "/")
-
-  # Run and return
-  data <- pv_query(parameters, reformat, ...)
+  article <- gsub(x = article, pattern = " ", replacement = "_", fixed = TRUE)
```
Comment (Ironholds) [inline R/query.R:41]: escaping is about more than just spaces. This isn't an adequate replacement for full encoding.

File: R/utils.R [hunk 2]
```diff
@@ -1,11 +1,20 @@
 #'@importFrom httr stop_for_status GET user_agent content status_code
 pv_query <- function(params, reformat = TRUE, ...){
+  if(length(params) > 1){
+    data <- lapply(as.list(params), pv_query, reformat = reformat)
+    if(reformat == TRUE){
+      return(do.call(rbind, data))
+    } else {
+      return(data)
```
Comment (Ironholds) [inline R/utils.R]: This seems pretty confusing in practice. Absent reformatting the structure is suddenly an extra degree nested. Documentation will need to be a lot more expansive and I'm not sure how I feel about the change.

File: R/query.R [hunk 3]
```diff
@@ -34,33 +34,18 @@
...
+article_pageviews <- function(project = "en.wikipedia", article = "R (programming language)"
+                              , platform = "all", user_type = "all", granularity = "daily"
+                              , start = "2015100100", end = NULL, reformat = TRUE, ...){
 
-
-  article <- gsub(x = article, pattern = " ", replacement = "_", fixed = TRUE)
-  if(length(article) > 1){
-    return(lapply(article, article_pageviews, project = project, platform = platform,
-                  user_type = user_type, start = start, end = end, ...))
+  if(granularity != "daily"){
```
Comment (Ironholds) [inline R/query.R]: Previously discussed.

**Comment by jeremiahpslewis** [comment]
Escaping was still there, but hidden in the pageviews function. I've moved it up to the article_pageviews function, where it belongs.

File: R/utils.R (hunk 2 above)
Comment (jeremiahpslewis) [inline R/utils.R]: I took your code as the model for this line (I don't tend to use too much recursion in R) This is just a further abstraction of your code in the article query:

```
  if(length(article) > 1){
    return(lapply(article, article_pageviews, project = project, platform = platform,
                  user_type = user_type, start = start, end = end, ...))
  }
```

Comment (jeremiahpslewis) [inline R/utils.R]: Alternately we could put this code in a function, `pv_query`, and create a separate function `pv_query_single` for a single query, which is not exported. Which ever way you prefer.

**Comment by Ironholds** [comment]
_thumbs up_

**Comment by jeremiahpslewis** [comment]
Added the granularity tweak. Last open issue is the lapply statement, right?

**Comment by jeremiahpslewis** [comment]
Thanks for setting up the travis-ci!

**Comment by Ironholds** [comment]
Yeah, the lapply statement and the doc fails in travis

**Comment by jeremiahpslewis** [comment]
on it...

**Comment by jeremiahpslewis** [comment]
Let me know about the lapply question...I've mocked up a different approach in the lapply branch on my fork.

File: R/utils.R (hunk 2 above)
Comment (Ironholds) [inline R/utils.R]: Fair! Hmn. I mean, noodling on it if we can assume reformatting by default, I'm fine with it.

Comment (jeremiahpslewis) [inline R/utils.R]: Reformatting is not assumed...if reformatting is false, then a list of lists will be returned where each raw response is a list element. Each data frame uniquely defines what the views refer to, within the context of each api. As a result, as long as reformat is true, it should not be an issue to rbind the results. :)

Comment (Ironholds) [inline R/utils.R]: Sure, that's what I meant; TRUE by default == assumed!

Only other bits:
1. Why is there a default for the reformat argument in pv_query? It's not exposed, afaik?
2. The vignette and docs will need updating to reflect the tweaks in output format/support for multiple params.

**Comment by jeremiahpslewis** [comment]
Ok, so vignettes are now updated. Fix is included with `pv_query` and `pv_query_single` for added clarity. Tests added to type-check function responses (and fixes for a few type issues that emerged).

File: R/utils.R (hunk 2 above)
Comment (jeremiahpslewis) [inline R/utils.R]: Fixed 1 & 2.

**Comment by jeremiahpslewis** [comment]
@Ironholds: Hey! Can this be merged or are their tweaks you would still like made?

**Comment by Ironholds** [comment]
Heya; sorry, I've been travelling all day. I'll give it another look.

File: R/query.R [hunk 4]
```diff
@@ -34,33 +34,15 @@
 #'obama_pageviews <- article_pageviews(article = "Barack_Obama")
 #'
 #'@export
-article_pageviews <- function(project = "en.wikipedia", article = "R (programming language)",
-                              platform = "all", user_type = "all",
-                              start = "2015100100", end = NULL, reformat = TRUE, ...){
-
+article_pageviews <- function(project = "en.wikipedia", article = "R (programming language)"
+                              , platform = "all", user_type = "all"
```
Comment (Ironholds) [inline R/query.R]: can you put the commas at the end of the previous line for stylistic purposes?

File: R/utils.R [hunk 5]
```diff
@@ -1,11 +1,22 @@
 #'@importFrom httr stop_for_status GET user_agent content status_code
-pv_query <- function(params, reformat = TRUE, ...){
+pv_query <- function(params, reformat, ...){
+  # Run multiple queries, return as list of results
+  data_list <- lapply(as.list(params), pv_query_single, reformat = reformat)
+
+  if(reformat == TRUE){ # collapses all results into one dataframe
+    return(do.call(rbind, data_list))
+  } else {
+    return(data_list)
```
Comment (Ironholds) [inline R/utils.R]: you don't need an else here; if it gets to that statement, the statement isn't rue. If it were the function would've returned and terminated.

File: R/utils.R [hunk 6]
```diff
@@ -35,22 +47,29 @@ reformat_data <- function(data){
...
-    data$date <- as.Date(paste(data$year, data$month, data$day, sep = "-"))
+    data$date <- as.POSIXct(paste(data$year, data$month, data$day, sep = "-"))
     data$rank <- as.numeric(data$rank)
-
   } else {
-    data <- data.frame(do.call(rbind, data), stringsAsFactors = FALSE)
-
-    data$date <- as.Date(substr(data$timestamp, 1, 8), format = "%Y%m%d")
+    data <- as.data.frame(do.call(rbind, data), stringsAsFactors = FALSE)
+    data$date <- as.POSIXct(as.character(data$timestamp), format = "%Y%m%d%H", tz = "GMT")
```
Comment (Ironholds) [inline R/utils.R:53]: Why the full POSIX timestamps? It can only go on a daily basis, no?

File: R/utils.R [hunk 7]
```diff
@@ -35,22 +47,29 @@ reformat_data <- function(data){
...
 
+  # Convert all variables to appropriate classes & formats.
   data$views <- as.numeric(data$views)
   data$language <- gsub("(.*)\\.(.*)", "\\1", data$project)
   data$project <- gsub("(.*)\\.(.*)", "\\2", data$project)
+  data$access <- as.character(data$access)
+  data$granularity <- as.character(data$granularity)
+
+  # Agent and article only contained in some queries
+  if(!is.null(data$agent)) data$agent <- as.character(data$agent)
```
Comment (Ironholds) [inline R/utils.R]: can you split these into multiline conditionals as the rest are?

File: man/project_pageviews.Rd [hunk 8]
```diff
@@ -12,10 +12,10 @@ project_pageviews(project = "en.wikipedia", platform = "all",
 \item{project}{the name of the project, structured as \code{[language_code].[project]}
 (see the default).}
 
-\item{platform}{The platform the pageviews came from; one of "all", "desktop", "mobile-web" and
+\item{platform}{The platform the pageviews came from; vector of "all", "desktop", "mobile-web" and
```
Comment (Ironholds) [inline man/project_pageviews.Rd]: "One or more of". "vector" leaves it ambiguous. Ditto below.

File: man/top_articles.Rd [hunk 9]
```diff
@@ -4,22 +4,21 @@
 \alias{top_articles}
 \title{Retrieve Data on Top Articles}
 \usage{
-top_articles(project = "en.wikipedia", platform = "all", year = "2015",
-  month = "10", day = "01", reformat = TRUE, ...)
+top_articles(project = "en.wikipedia", platform = "all",
+  start = as.Date("2015-10-01"), granularity = "daily", reformat = TRUE,
```
Comment (Ironholds) [inline man/top_articles.Rd:8]: why the ...?

File: R/query.R (hunk 4 above)
Comment (jeremiahpslewis) [inline R/query.R]: :+1:

File: R/utils.R (hunk 5 above)
Comment (jeremiahpslewis) [inline R/utils.R]: :+1:

File: R/utils.R (hunk 6 above)
Comment (jeremiahpslewis) [inline R/utils.R:53]: Yes, I can change this if you would like, but I would prefer to return a consistent date type across all package functions.

File: R/utils.R (hunk 7 above)
Comment (jeremiahpslewis) [inline R/utils.R]: :+1:

File: man/project_pageviews.Rd (hunk 8 above)
Comment (jeremiahpslewis) [inline man/project_pageviews.Rd]: :+1:

File: man/top_articles.Rd (hunk 9 above)
Comment (jeremiahpslewis) [inline man/top_articles.Rd:8]: Lets the user use R date objects without worrying about parsing / conforming to a unique api. Also keeps parameters consistent across all functions.

Comment (jeremiahpslewis) [inline man/top_articles.Rd:8]: Part of the usefulness of an API wrapper should be that it allows R users to keep things as far as possible in a consistent, R vernacular and leverage existing data types.

File: R/utils.R (hunk 6 above)
Comment (Ironholds) [inline R/utils.R:53]: Agreed. Are there any API calls that generate anything more granular than daily right now?

File: man/top_articles.Rd (hunk 9 above)
Comment (Ironholds) [inline man/top_articles.Rd:8]: Wait, yeah, I misread where it was being used. Whoops :D

File: R/utils.R (hunk 6 above)
Comment (jeremiahpslewis) [inline R/utils.R:53]: Yep, for the project stats.

Comment (Ironholds) [inline R/utils.R:53]: _thumbs up_

**Comment by Ironholds** [comment]
We're done :D Thanks so much; sorry it took so long!

**Comment by jeremiahpslewis** [comment]
Woot!

**Pull request merged**